ignite/contrib/handlers/ema_handler.py

```diff
@@ -0,0 +1,204 @@
+        assert 0 < momentum_warmup < 1
+        assert momentum_warmup <= momentum
+        assert isinstance(interval, int) and interval > 0
+        assert isinstance(warmup_iters, int) and warmup_iters > 0
+        assert isinstance(model, nn.Module)
+        self.momentum = momentum
+        self.interval = interval
+        self.momentum_warmup = momentum_warmup
+        self.warmup_iters = warmup_iters
+        self.device = device
+
+        self.ema = deepcopy(self._unwrap_model(model))
+        self.ema.eval()
+        if self.device is not None:
+            self.ema.to(device=device)
+
+    def state_dict(self, **kwargs: Optional[Any]) -> OrderedDict[str, Tensor]:
+        """Return ``state_dict`` of the EMA copy.
+
+        Args:
+            kwargs: arguments of ``nn.Module.state_dict``.
+
+        """
+        return self.ema.state_dict(**kwargs)
+
+    def load_state_dict(self, state_dict: OrderedDict[str, Tensor], strict: bool = True) -> None:
+        """Load the ``sate_dict`` to the EMA copy.
+
+        Args:
+            state_dict: dict contains the weights.
+            strict: whether to strictly enforce that the keys match.
+
+        """
+        self.ema.load_state_dict(state_dict, strict=strict)
+
+    @staticmethod
+    def _unwrap_model(model: nn.Module) -> nn.Module:
+        if isinstance(model, (nn.parallel.DataParallel, nn.parallel.DistributedDataParallel)):
+            return model.module
+        else:
+            return model
+
+    def _get_momentum(self, curr_iter: int):
+        """Get current momentum, `curr_iter` should be 1-based. When `curr_iter = 1`, `momentum =
+        self.momentum_warmup`; when `curr_iter >= self.warmup_iters`, `momentum = self.momentum`"""
+        assert curr_iter >= 1
+        denominator = max(1, self.warmup_iters - 1)
+        momentum = self.momentum_warmup + (self.momentum - self.momentum_warmup) * (curr_iter - 1) / denominator
+        return min(self.momentum, momentum)
+
+    @torch.no_grad()
+    def _swap_params(self, engine: Engine, model: nn.Module) -> None:
+
+        model = self._unwrap_model(model)
+        for ema_v, model_v in zip(self.ema.state_dict().values(), model.state_dict().values()):
+            if self.device is not None:
+                model_v = model_v.to(device=self.device)
+            tmp = model_v.data.clone()
+            model_v.data.copy_(ema_v.data)
+            ema_v.data.copy_(tmp)
+
+    @torch.no_grad()
+    def _update(self, engine: Engine, model: nn.Module) -> None:
+
+        model = self._unwrap_model(model)
+        curr_iter = engine.state.iteration
+        momentum = self._get_momentum(curr_iter)
+
+        for ema_v, model_v in zip(self.ema.state_dict().values(), model.state_dict().values()):
+            if self.device is not None:
+                model_v = model_v.to(device=self.device)
+            ema_v.mul_(1 - momentum).add_(model_v.data, alpha=momentum)
+
+    def attach(self, engine: Engine, model: nn.Module) -> None:
```

Comment:
I see that we ask the model to EMA twice: in the constructor and here. I think it would be better to do that once and maybe in the constructor...

Reply:
Yes, I updated it in the next commit.